### PR TITLE
V2 filters backend

### LINKED ISSUE #70: Advanced Unit Browsing - V2 Filters [ Epic 🏔️ ]
# Overview

The second generation of USHS's state-of-the-art unit filtering interface

[Figma](https://www.figma.com/file/06vEDN5gtlHlC5hxNXMb4g/USHS-Designs?node-id=1935%3A11952&mode=dev)

## Tasks
- [x] #80 
- [x] #100

### LINKED ISSUE #100: V2 Filters Backend Integration
# Overview
integrate the frontend of the v2 filters (#80) with the backend of the project

## Frontend
Make some finishing touches to the frontend:
- Remove the reset filter button
- Make sure the "Clear All" button reset all the filters including the search and sort
- Finish up the filter api route in the frontend
- Add filters as a context (useContext) and refactor code

## Backend
Implement the new filters in the backend
- Security
- Application Fee
- Size
- Date available
- Housing authority
- Accessibility
- Rental Criteria
- Additional Rules

## Acceptance Criteria
- [ ] Finishing touches in the frontend
- [ ] Backend integration

### PR DESCRIPTION
**Pull request by RamtinTJB**
## Tracking Info

Resolves #100 
Resolves #70 

## Changes

- Added new fields for the new filters in the frontend and backend
- Integrated the v2 filters in the backend API
- Refactored frontend code
- Added FiltersConext

## Testing

Manually verified the different combination of filters to make sure they display the right results

**Note**: I haven't figured out a way to implement the multi-option (checkbox) yet I'll add that pretty soon.

## Confirmation of Change

image

### CODE REVIEW

File: backend/src/services/units.ts [hunk 1]
```diff
@@ -115,11 +144,25 @@ export const getUnits = async (filters: FilterParams) => {
       break;
   }
 
+  //const hasAccessibility = !(filters.accessibility === undefined || filters.accessibility === "[]");
+  //const rentalCriteria = !(filters.rentalCriteria === undefined || filters.rentalCriteria === "[]");
+  // const additionalRules = !(
+  //   filters.additionalRules === undefined || filters.additionalRules === "[]"
+  // );
+
```
Comment (petabite) [inline backend/src/services/units.ts:152]: Can we remove this?

Comment (petabite) [inline backend/src/services/units.ts]: Are these filters not working?

File: frontend/src/api/units.ts [hunk 2]
```diff
@@ -67,6 +79,7 @@ export async function getUnits(params: FilterParams): Promise<APIResult<Unit[]>>
   try {
     const queryParams = new URLSearchParams(params);
     const url = `/units?${queryParams.toString()}`;
+    console.log(url);
```
Comment (petabite) [inline frontend/src/api/units.ts]:
```suggestion
```

File: frontend/src/components/FilterPanel.tsx [hunk 3]
```diff
@@ -118,25 +122,127 @@ export const FitlerPanel = () => {
...
+      maxSize: sizeState.max.toString(),
+      fromDate: dateState.from,
+      toDate: dateState.to,
+    };
+
+    setFilters({
+      ...newFilters,
+      approved: filters.approved,
+      sort: filters.sort ?? "",
+      search: filters.search ?? "0",
```
Comment (petabite) [inline frontend/src/components/FilterPanel.tsx]:
```suggestion
      search: filters.search ?? "",
```

There seems to be a bug where clicking "Apply" sets the search value to "0"

File: frontend/src/components/MinMaxFilter.tsx [hunk 4]
```diff
@@ -1,16 +1,16 @@
...
-import { FilterRangeInput } from "./FilterRangeInput";
+import { FilterRangeInput, FilterRangeInputProps } from "./FilterRangeInput";
 
-export type MinMaxFilterProps = {
+export type MinMaxFilterProps = FilterRangeInputProps & {
   title: string;
 };
 
-export const MinMaxFitler = (props: MinMaxFilterProps) => {
+export const MinMaxFilter = (props: MinMaxFilterProps) => {
```
Comment (petabite) [inline frontend/src/components/MinMaxFilter.tsx:9]: lol nice

File: frontend/src/pages/Home.tsx [hunk 5]
```diff
@@ -51,32 +54,30 @@ export function Home() {
...
-            refreshUnits={(approved) => {
-              const newFilters = { ...filters, approved };
-              fetchUnits(newFilters);
-              setFilters(newFilters);
-            }}
-          />
-        </div>
-      </HomePageLayout>
-    </Page>
+    <FiltersContext.Provider value={{ filters, setFilters }}>
```
Comment (petabite) [inline frontend/src/pages/Home.tsx:89]: It also looks like the filter context stuff might be broken. When I click "Back" from the Unit Details page, the filter criteria get reset to their default values.

**Review by petabite — CHANGES_REQUESTED** [review]
Looking great 🔥 💯   Left a couple comments below.

Also, do you think it's possible to implement the "$" and "sqft" prefix/postfix as in the designs?

image

File: frontend/src/pages/Home.tsx (hunk 5 above)
Comment (petabite) [reply frontend/src/pages/Home.tsx:89]: Ah, I did some digging and it looks like the `FilterContext.Provider` got removed in [this commit](https://github.com/TritonSE/USHS-Housing-Portal/commit/02ad2f7d90b5c11bef3d7b5bbc894b013d204660#diff-9f4fbb52c544c388744cd765e62f81f3049a5d32948140308a6a623ee4a01005).

**Review by petabite — COMMENTED** [review]
This looks awesome! I just added some code to support persisting the filters when the user clicks "Back" on the unit details page. Should be good to go once you add the "Any" option for the "Housing Authority" filter!

**Review by petabite — APPROVED** [review]
🚢 !!

**Review by soma-p — APPROVED** [review]
PR looks great!

**Pull request merged**